File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/utities/MapViewHandler.java
package vinhlong.ditagis.com.khaosatdongho.utities;

import android.app.Activity;
import android.view.MotionEvent;
import android.widget.Toast;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.ArcGISFeature;
import com.esri.arcgisruntime.data.Feature;
import com.esri.arcgisruntime.data.FeatureQueryResult;
import com.esri.arcgisruntime.data.Field;
import com.esri.arcgisruntime.data.QueryParameters;
import com.esri.arcgisruntime.data.ServiceFeatureTable;
import com.esri.arcgisruntime.geometry.Geometry;
import com.esri.arcgisruntime.geometry.GeometryEngine;
import com.esri.arcgisruntime.geometry.Point;
import com.esri.arcgisruntime.geometry.SpatialReferences;
import com.esri.arcgisruntime.mapping.Viewpoint;
import com.esri.arcgisruntime.mapping.view.MapView;

import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ExecutionException;

import vinhlong.ditagis.com.khaosatdongho.MainActivity;
import vinhlong.ditagis.com.khaosatdongho.adapter.DanhSachDongHoKHAdapter;
import vinhlong.ditagis.com.khaosatdongho.async.EditGeometryAsync;
import vinhlong.ditagis.com.khaosatdongho.async.SingleTapAdddFeatureAsync;
import vinhlong.ditagis.com.khaosatdongho.async.SingleTapMapViewAsync;
import vinhlong.ditagis.com.khaosatdongho.entities.DApplication;


/**
 * Created by ThanLe on 2/2/2018.
 */

public class MapViewHandler extends Activity {
    private MapView mapView;
    private ServiceFeatureTable dongHoKHSFT;
    private Popup popup;
    private MainActivity mainActivity;
    private DApplication dApplication;

    public MapViewHandler(MapView mapView, MainActivity mainActivity, Popup popup) {
        this.mapView = mapView;
        this.mainActivity = mainActivity;
        this.popup = popup;
        this.dApplication = (DApplication) mainActivity.getApplication();
    }

    public void setDongHoKHSFT(ServiceFeatureTable dongHoKHSFT) {
        this.dongHoKHSFT = dongHoKHSFT;
    }

    public void addFeature() {
        Point add_point = mapView.getCurrentViewpoint(Viewpoint.Type.CENTER_AND_SCALE).getTargetGeometry().getExtent().getCenter();
        new SingleTapAdddFeatureAsync(mainActivity, mapView, this.dongHoKHSFT, feature -> {
//            this.dApplication.getMainActivity().dismissPin();
            if (feature != null) {
                this.popup.showPopup(feature);
            } else this.popup.dimissCallout();
        }).execute(add_point);

    }

    public void updateGeometry(ArcGISFeature feature) {
        Point point = mapView.getCurrentViewpoint(Viewpoint.Type.CENTER_AND_SCALE).getTargetGeometry().getExtent().getCenter();

        //Kiểm tra cùng ngày, cùng vị trí đã có sự cố nào chưa, nếu có thì cảnh báo, chưa thì thêm bình thường
        new EditGeometryAsync(mainActivity, this.dongHoKHSFT, feature, aBoolean -> {
            if (aBoolean != null && aBoolean) {
                mainActivity.setChangingGeometry(false,null);

                Toast.makeText(mainActivity, "Đổi vị trí thành công", Toast.LENGTH_SHORT).show();
            }
            else {
                Toast.makeText(mainActivity, "Có lỗi xảy ra", Toast.LENGTH_SHORT).show();
            }
        }).execute(point);
    }
    public double[] onScroll() {
        Point center = mapView.getCurrentViewpoint(Viewpoint.Type.CENTER_AND_SCALE).getTargetGeometry().getExtent().getCenter();
        Geometry project = GeometryEngine.project(center, SpatialReferences.getWgs84());
        return new double[]{project.getExtent().getCenter().getX(), project.getExtent().getCenter().getY()};
    }

    public void onSingleTapMapView(MotionEvent e) {
        android.graphics.Point point = new android.graphics.Point((int) e.getX(), (int) e.getY());
        new SingleTapMapViewAsync(mainActivity, mapView, feature -> {
            if (feature != null) {
                this.popup.showPopup(feature);
            } else this.popup.dimissCallout();
        }).execute(point);

    }

    public void queryByObjectID(String objectID) {
        final QueryParameters queryParameters = new QueryParameters();
        final String query = "OBJECTID = " + objectID;
        queryParameters.setWhereClause(query);
        final ListenableFuture<FeatureQueryResult> feature = dongHoKHSFT.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL);
        feature.addDoneListener(() -> {
            try {
                FeatureQueryResult result = feature.get();
                if (result.iterator().hasNext()) {
                    Feature item = result.iterator().next();
                    showPopup(item);
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (ExecutionException e) {
                e.printStackTrace();
            }
        });

    }

    private void showPopup(Feature selectedFeature) {
        if (selectedFeature != null) {
            popup.showPopup((ArcGISFeature) selectedFeature);
        }
    }

    public void querySearch(String searchStr, final DanhSachDongHoKHAdapter adapter) {
        adapter.clear();
        adapter.notifyDataSetChanged();
        QueryParameters queryParameters = new QueryParameters();
        StringBuilder builder = new StringBuilder();
        for (Field field : dongHoKHSFT.getFields()) {
            switch (field.getFieldType()) {
                case OID:
                case INTEGER:
                case SHORT:
                    try {
                        int search = Integer.parseInt(searchStr);
                        builder.append(String.format("%s = %s", field.getName(), search));
                        builder.append(" or ");
                    } catch (Exception e) {

                    }
                    break;
                case FLOAT:
                case DOUBLE:
                    try {
                        double search = Double.parseDouble(searchStr);
                        builder.append(String.format("%s = %s", field.getName(), search));
                        builder.append(" or ");
                    } catch (Exception e) {

                    }
                    break;
                case TEXT:
                    builder.append(field.getName() + " like N'%" + searchStr + "%'");
                    builder.append(" or ");
                    break;
            }
        }
        builder.append(" 1 = 2 ");
        queryParameters.setWhereClause(builder.toString());
        queryParameters.setMaxFeatures(100);
        final ListenableFuture<FeatureQueryResult> feature = dongHoKHSFT.queryFeaturesAsync(queryParameters);
        feature.addDoneListener(() -> {
            try {
                FeatureQueryResult result = feature.get();
                Iterator iterator = result.iterator();
                while (iterator.hasNext()) {
                    Feature item = (Feature) iterator.next();
                    Map<String, Object> attributes = item.getAttributes();
                    DanhSachDongHoKHAdapter.Item dongHoKH = new DanhSachDongHoKHAdapter.Item();
                    dongHoKH.setObjectID(attributes.get(Constant.LayerFields.OBJECTID).toString());
                    Object danhboDongHo = attributes.get(Constant.DongHoKhachHangFields.ID);
                    Object tenKH = attributes.get(Constant.DongHoKhachHangFields.TenKH);
                    Object maKhachHang = attributes.get(Constant.DongHoKhachHangFields.MaKhachHang);
                    if (danhboDongHo != null)
                        dongHoKH.setIdDongHo(danhboDongHo.toString());
                    if (tenKH != null) {
                        dongHoKH.setTenKhachHang(tenKH.toString());
                    }
                    if (maKhachHang != null) {
                        dongHoKH.setMaKhachHang(maKhachHang.toString());
                    }
                    adapter.add(dongHoKH);
                    adapter.notifyDataSetChanged();
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (ExecutionException e) {
                e.printStackTrace();
            }
        });

    }
}

File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/utities/MapViewHandler.kt
package vinhlong.ditagis.com.khaosatdongho.utities

import android.app.Activity
import android.view.MotionEvent
import android.widget.Toast
import com.esri.arcgisruntime.data.*
import com.esri.arcgisruntime.geometry.GeometryEngine
import com.esri.arcgisruntime.geometry.SpatialReferences
import com.esri.arcgisruntime.mapping.Viewpoint
import com.esri.arcgisruntime.mapping.view.MapView
import kotlinx.android.synthetic.main.activity_main.*
import vinhlong.ditagis.com.khaosatdongho.MainActivity
import vinhlong.ditagis.com.khaosatdongho.adapter.DanhSachDongHoKHAdapter
import vinhlong.ditagis.com.khaosatdongho.async.EditGeometryAsync
import vinhlong.ditagis.com.khaosatdongho.async.SingleTapAddFeatureAsync
import vinhlong.ditagis.com.khaosatdongho.async.SingleTapMapViewAsync
import vinhlong.ditagis.com.khaosatdongho.entities.DApplication
import java.util.concurrent.ExecutionException


/**
 * Created by ThanLe on 2/2/2018.
 */

class MapViewHandler(private val mapView: MapView, private val mActivity: MainActivity, private val popup: Popup) : Activity() {
    private var dongHoKHSFT: ServiceFeatureTable? = null
    private val mApplication: DApplication = mActivity.application as DApplication

    fun setDongHoKHSFT(dongHoKHSFT: ServiceFeatureTable) {
        this.dongHoKHSFT = dongHoKHSFT
    }

    fun addFeature() {
        val add_point = mapView.getCurrentViewpoint(Viewpoint.Type.CENTER_AND_SCALE).targetGeometry.extent.center
        mApplication.progressDialog.show(mActivity, mActivity.container_main, "Đang thêm đồng hồ khách hàng...")
        SingleTapAddFeatureAsync(mActivity, mapView, this.dongHoKHSFT!!, object : SingleTapAddFeatureAsync.AsyncResponse {
            override fun processFinish(feature: ArcGISFeature?) {
                if (feature != null) {
                    this@MapViewHandler.popup.showPopup(feature)
                } else
                    this@MapViewHandler.popup.dimissCallout()
                mApplication.progressDialog.dismiss()
            }

            //            this.mApplication.getMainActivity().dismissPin();

        }).execute(add_point)

    }

    fun updateGeometry(feature: ArcGISFeature) {
        val point = mapView.getCurrentViewpoint(Viewpoint.Type.CENTER_AND_SCALE).targetGeometry.extent.center
        mApplication.progressDialog.show(mActivity, mActivity.container_main, "Đang cập nhật tọa độ...")
        //Kiểm tra cùng ngày, cùng vị trí đã có sự cố nào chưa, nếu có thì cảnh báo, chưa thì thêm bình thường
        EditGeometryAsync(mActivity, this.dongHoKHSFT!!, feature, object : EditGeometryAsync.AsyncResponse {
            override fun processFinish(isSuccess: Boolean?) {
                if (isSuccess != null && isSuccess) {
                    mActivity.setChangingGeometry(false, null)

                    Toast.makeText(mActivity, "Đổi vị trí thành công", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(mActivity, "Có lỗi xảy ra", Toast.LENGTH_SHORT).show()
                }
                mApplication.progressDialog.dismiss()
            }

        }).execute(point)
    }

    fun onScroll(): DoubleArray {
        val center = mapView.getCurrentViewpoint(Viewpoint.Type.CENTER_AND_SCALE).targetGeometry.extent.center
        val project = GeometryEngine.project(center, SpatialReferences.getWgs84())
        return doubleArrayOf(project.extent.center.x, project.extent.center.y)
    }

    fun onSingleTapMapView(e: MotionEvent) {
        val point = android.graphics.Point(e.x.toInt(), e.y.toInt())
        mApplication.progressDialog.show(mActivity, mActivity.container_main, "Đang xác định đồng hồ khách hàng...")
        SingleTapMapViewAsync(mActivity, mapView, object : SingleTapMapViewAsync.AsyncResponse {
            override fun processFinish(feature: ArcGISFeature?) {
                mApplication.progressDialog.dismiss()
                if (feature != null) {
                    this@MapViewHandler.popup.showPopup(feature)
                } else
                    this@MapViewHandler.popup.dimissCallout()
            }


        }).execute(point)

    }


    fun showPopup(selectedFeature: Feature?) {
        if (selectedFeature != null) {
            popup.showPopup(selectedFeature as ArcGISFeature)
        }
    }

    fun querySearch(searchStr: String, adapter: DanhSachDongHoKHAdapter) {
        adapter.clear()
        adapter.notifyDataSetChanged()
        val queryParameters = QueryParameters()
        val builder = StringBuilder()
        for (field in dongHoKHSFT!!.fields) {
            when (field.fieldType) {
                Field.Type.OID, Field.Type.INTEGER, Field.Type.SHORT -> try {
                    val search = Integer.parseInt(searchStr)
                    builder.append(String.format("%s = %s", field.name, search))
                    builder.append(" or ")
                } catch (e: Exception) {

                }

                Field.Type.FLOAT, Field.Type.DOUBLE -> try {
                    val search = java.lang.Double.parseDouble(searchStr)
                    builder.append(String.format("%s = %s", field.name, search))
                    builder.append(" or ")
                } catch (e: Exception) {

                }

                Field.Type.TEXT -> {
                    builder.append(field.name + " like N'%" + searchStr + "%'")
                    builder.append(" or ")
                }
            }
        }
        builder.append(" 1 = 2 ")
        queryParameters.whereClause = builder.toString()
        queryParameters.maxFeatures = 100
        val featureQueryResultListenableFuture = dongHoKHSFT!!.queryFeaturesAsync(queryParameters)
        featureQueryResultListenableFuture.addDoneListener {
            try {
                val result = featureQueryResultListenableFuture.get()
                val iterator = result.iterator()
                while (iterator.hasNext()) {
                    val feature = iterator.next() as Feature

                    adapter.add(feature)
                    adapter.notifyDataSetChanged()
                }
            } catch (e: InterruptedException) {
                e.printStackTrace()
            } catch (e: ExecutionException) {
                e.printStackTrace()
            }
        }

    }
}